---
title: "Targeted enrichment and discovery of jumbo phages"
author: "Jeffrey Blanchard"
date: "5/4/2023"
output:
  html_document:
    toc: true
    toc_depth: 4
    theme: united
    highlight: tango
---

# Overview

This report is an analysis of the Barre Woods filter metagenomes from the combined assembly  Assembly, binning, quality control and classification was done on KBase
(https://narrative.kbase.us/narrative/145971. The checkv and virsorter files were downloaded to this project folder.


```{r}
# Load libraries
library(tidyverse)
library(DT)
```

## virsorter2 and vcheck
### Data processing
virsorter2 identifies dsDNAphages, NCLDV, ssDNA, virophages and some rna viruses in contigs. An 0.8 cutoff score is commonly used to remove potential false positives. checkv assesses the quality of contigs.

```{r}
virsorter <- read_tsv("data_combined_assembly/virsorter-final-viral-score.tsv") %>% 
  filter(max_score_group == "dsDNAphage") %>% 
  select(-c(dsDNAphage, ssDNA, NCLDV, lavidaviridae)) %>% 
  rename("contig_id" = "seqname")

checkv <- read_tsv("data_combined_assembly/checkv_quality_summary.tsv") 
```

#### Join checkv and virsorter dataframes

```{r}
virsorter_checkv <- left_join(virsorter, checkv, by = "contig_id")
```

#### filter to high quality and complete genomes

```{r}
virsorter_checkv_HQ <- virsorter_checkv %>% 
  filter(checkv_quality == "Complete" | checkv_quality == "High-quality")
```

#### filter to HQ jumbo phages

```{r}
virsorter_checkv_jumbo <- virsorter_checkv %>% 
  filter(checkv_quality == "Complete" | checkv_quality == "High-quality") %>% 
  filter(length >= 200000)
```

#### Write files
```{r, eval=FALSE}
# write summary file
write_tsv(virsorter_checkv, "data_combined_assembly/coa-virsorter_checkv.tsv") 
write_tsv(virsorter_checkv_HQ, "data_combined_assembly/coa-virsorter_checkv_HQ.tsv") 
write_tsv(virsorter_checkv_jumbo, "data_combined_assembly/coa-virsorter_checkv_jumbo.tsv") 
```

### Data analysis

#### Barchart of checkv_quality
```{r}
virsorter_checkv %>% 
ggplot(aes(x = checkv_quality)) + 
  geom_bar() 
```

#### Table with % checkv_quality in each category
```{r}
virsorter_checkv %>% 
  mutate(across(checkv_quality, as_factor)) %>% 
  group_by(checkv_quality) %>% 
  summarise(n = n()) %>%
  mutate(freq = n / sum(n))
```

#### Histogram of all contigs
Most low quality contigs are less than 20kb
```{r}
virsorter_checkv %>% 
ggplot(aes(x = length, fill = checkv_quality)) + 
  geom_histogram(colour = "black",  binwidth=10000) +
  ggtitle("Genome size of Phage") +
  xlab("Genome size") + 
  theme(text = element_text(size = 20, color="black"))
  theme(axis.text.x = element_text(angle = 90)) 
```
#### Histogram of HQ contigs
```{r}
virsorter_checkv_HQ %>% 
ggplot(aes(x = length, fill = checkv_quality)) + 
  geom_histogram(colour = "black",  binwidth=10000) +
  ggtitle("Genome size of Phage") +
  xlab("Genome size") + 
  #theme(text = element_text(size = 20, color="black"))
  theme(axis.text.x = element_text(angle = 45)) 
```

## vContact
vContact groups contigs into clusters along with phages in NCBI's viurs reference database
### data  processing

#### read in files
```{r}
vcontact <- read_csv("data_combined_assembly/vcontact_genome_by_genome_overview.csv") 
vcontact_ntw <- read_table("data_combined_assembly/coa-vcontact_c1.ntw", col_names = FALSE) 
```

#### join with virsorter and checkv results
```{r}
virsorter_checkv_HQ_vcontact <- left_join(virsorter_checkv_HQ, vcontact, by = c("contig_id" = "Genome"))
virsorter_checkv_jumbo_vcontact <- left_join(virsorter_checkv_jumbo, vcontact, by = c("contig_id" = "Genome"))
```

#### filter network files to contain just HQ or jumbo phages for cytoscape
```{r}
# make a vector of HQ and jumbo contigs
HQ_contigs <- pull(virsorter_checkv_HQ, contig_id)
jumbo_contigs <- pull(virsorter_checkv_jumbo_vcontact, contig_id)  

# create list of low quality genomes
# There are genes in the vcontact network that are not in the merged virsorter/vcheck file
# So first make a list of vcontact genomes with contig in name then filter HQ

vcontact_contig_LQ <- vcontact %>% 
filter(grepl('Contig', Genome)) %>% 
filter(!Genome %in% HQ_contigs)

# make a vector of low quality contigs 
LQ_contigs <- pull(vcontact_contig_LQ, Genome)  

# filter vcontact network based on HQ genomes}
vcontact_ntw_HQ <- vcontact_ntw %>% 
filter(X1 %in% HQ_contigs) %>% 
filter(!X1 %in% LQ_contigs & !X2 %in% LQ_contigs)

# filter vcontact network based on jumbo genomes
vcontact_ntw_jumbo <- vcontact_ntw %>% 
filter(X1 %in% jumbo_contigs) %>% 
filter(!X1 %in% LQ_contigs & !X2 %in% LQ_contigs)

# filter vcontact network based on jumbo genomes and interaction score <30
vcontact_ntw_jumbo30 <- vcontact_ntw_jumbo %>% 
  filter(X3 > 30)
```

#### Write files
```{r, eval=FALSE}
write_tsv(virsorter_checkv_HQ_vcontact, "data_combined_assembly/coa-virsorter_checkv_HQ_vcontact.tsv") 
write_tsv(virsorter_checkv_jumbo_vcontact,"data_combined_assembly/coa-virsorter_checkv_jumbo_vcontact.tsv") 
write.table(vcontact_ntw_HQ, "data_combined_assembly/vcontact_ntw_HQ.ntw", sep = " ", col.names=F, row.names=F, quote=F)
write.table(vcontact_ntw_jumbo, "data_combined_assembly/vcontact_ntw_jumbo.ntw", sep = " ", col.names=F, row.names=F, quote=F)
write.table(vcontact_ntw_jumbo30, "data_combined_assembly/vcontact_ntw_jumbo30.ntw", sep = " ", col.names=F, row.names=F, quote=F)
```

### Data Analysis
#### Histogram of interaction score
```{r}
vcontact_ntw_jumbo %>% 
ggplot(aes(x = X3)) + 
  geom_histogram(colour = "black",  binwidth=10) +
  xlab("Score") + 
  #theme(text = element_text(size = 20, color="black"))
  theme(axis.text.x = element_text(angle = 45)) 
```


## Cytoscape
### Notes
#### To load network
File > Import> Network from file
Open
Advanced options
Select delimiter =  space
deselect use first column as names
select col 1 as source
select col 2 as target
select col 3 as edge

#### To select a group of nodes
shift then mouse drag

### Data Analysis
#### Jumbo30 network

![jumbo30](data_combined_assembly/jumbo30_cytoscape_ntw.png)

## DRAMv

### Data processing

#### Read files AMG summary and stats files

```{r}
DRAMv_amg <- read_tsv("data_combined_assembly/DRAMv_amg_summary.tsv") 
DRAMv_stats <- read_tsv("data_combined_assembly/DRAMv_vMAG_stats.tsv") %>% 
  rename("contig_id" = "...1")
```

#### join with stats virsorter and checkv results
```{r}
DRAMv_stats_virsorter_checkv_HQ <- left_join(virsorter_checkv_HQ, DRAMv_stats, by = "contig_id")
DRAMv_stats_virsorter_checkv_jumbo <- left_join(virsorter_checkv_jumbo, DRAMv_stats, by = "contig_id")
```

#### filter amd files to HQ and jumbo
```{r}
DRAMv_amg_HQ <- DRAMv_amg %>% 
filter(scaffold %in% HQ_contigs)

DRAMv_amg_jumbo <- DRAMv_amg %>% 
filter(scaffold %in% jumbo_contigs)
```

#### Write files

```{r}
write_tsv(DRAMv_stats_virsorter_checkv_HQ, "data_combined_assembly/coa-DRAMv_stats_virsorter_checkv_HQ .tsv") 
write_tsv(DRAMv_stats_virsorter_checkv_jumbo,"data_combined_assembly/coa-DRAMv_stats_virsorter_checkv_jumbo.tsv") 
write_tsv(DRAMv_amg_HQ, "data_combined_assembly/coa-DRAMv_amg_HQQ .tsv") 
write_tsv(DRAMv_amg_jumbo,"data_combined_assembly/coa-DRAMv_amg_jumbo.tsv") 
```

### Data Analysis

#### Graph of genes vs amgs in HQ

```{r}
DRAMv_stats_virsorter_checkv_HQ %>% 
ggplot(aes(x = `Gene count`, y = `potential AMG count`)) + 
  geom_point() 
```


#### Graph of genes vs amgs in HQ no provirus

```{r}
DRAMv_stats_virsorter_checkv_HQ %>% 
  filter(provirus == "No") %>% 
ggplot(aes(x = `Gene count`, y = `potential AMG count`)) + 
  geom_point() 
```
